### PR TITLE
ignore unexpected types instead of erroring out

### LINKED ISSUE #14146: Fatal error after updating to 3.88.0
### What happened?

After updating to 3.88.0 Pulumi CLI encounters fatal error

### Example

```
================================================================================
The Pulumi CLI encountered a fatal error. This is a bug!
We would appreciate a report: https://github.com/pulumi/pulumi/issues/
Please provide all of the below text in your report.
================================================================================
Pulumi Version:   v3.88.0
Go Version:       go1.21.1
Go Compiler:      gc
Architecture:     amd64
Operating System: linux
Panic:            fatal: A failure has occurred: unexpected wire type <nil>

goroutine 1 [running]:
runtime/debug.Stack()
        /opt/hostedtoolcache/go/1.21.1/x64/src/runtime/debug/stack.go:24 +0x5e
main.panicHandler(0xc00279ff07)
        /home/runner/work/pulumi/pulumi/pkg/cmd/pulumi/main.go:35 +0x45
panic({0x251a320?, 0xc002880090?})
        /opt/hostedtoolcache/go/1.21.1/x64/src/runtime/panic.go:914 +0x21f
gopkg.in/yaml%2ev3.handleErr(0xc00279e8b8)
        /home/runner/go/pkg/mod/gopkg.in/yaml.v3@v3.0.1/yaml.go:294 +0x6d
panic({0x251a320?, 0xc002880090?})
        /opt/hostedtoolcache/go/1.21.1/x64/src/runtime/panic.go:914 +0x21f
gopkg.in/yaml%2ev3.handleErr(0xc00279df60)
        /home/runner/go/pkg/mod/gopkg.in/yaml.v3@v3.0.1/yaml.go:294 +0x6d
panic({0x251a320?, 0xc002880090?})
        /opt/hostedtoolcache/go/1.21.1/x64/src/runtime/panic.go:914 +0x21f
github.com/pulumi/pulumi/sdk/v3/go/common/util/contract.failfast(...)
        /home/runner/work/pulumi/pulumi/sdk/go/common/util/contract/failfast.go:23
github.com/pulumi/pulumi/sdk/v3/go/common/util/contract.Failf({0x2ad658a?, 0x5?}, {0xc00279d570?, 0x41910b?, 0x2a?})
        /home/runner/work/pulumi/pulumi/sdk/go/common/util/contract/fail.go:32 +0xcb
github.com/pulumi/pulumi/sdk/v3/go/common/resource/config.unmarshalObject({0x0?, 0x0?})
        /home/runner/work/pulumi/pulumi/sdk/go/common/resource/config/object.go:492 +0x54a
github.com/pulumi/pulumi/sdk/v3/go/common/resource/config.unmarshalObject({0x2626a00?, 0xc0027b2450?})
        /home/runner/work/pulumi/pulumi/sdk/go/common/resource/config/object.go:468 +0x5b1
github.com/pulumi/pulumi/sdk/v3/go/common/resource/config.unmarshalObject({0x24e0e20?, 0xc00280c138?})
        /home/runner/work/pulumi/pulumi/sdk/go/common/resource/config/object.go:484 +0x6a5
github.com/pulumi/pulumi/sdk/v3/go/common/resource/config.(*object).UnmarshalYAML(0xc00279da30, 0xc0027b8560)
        /home/runner/work/pulumi/pulumi/sdk/go/common/resource/config/object.go:435 +0x5b
github.com/pulumi/pulumi/sdk/v3/go/common/resource/config.(*Value).UnmarshalYAML(0xc00280c018, 0xc00280c018?)
        /home/runner/work/pulumi/pulumi/sdk/go/common/resource/config/value.go:152 +0x28
gopkg.in/yaml%2ev3.(*decoder).callObsoleteUnmarshaler(0xc0009f5110, 0xc00256e460, {0x7f1b415620a8, 0xc00280c018})
        /home/runner/go/pkg/mod/gopkg.in/yaml.v3@v3.0.1/decode.go:379 +0x94
gopkg.in/yaml%2ev3.(*decoder).prepare(0xc00279dbc8?, 0x410593?, {0x2998d40?, 0xc00280c018?, 0x18?})
        /home/runner/go/pkg/mod/gopkg.in/yaml.v3@v3.0.1/decode.go:427 +0x21e
gopkg.in/yaml%2ev3.(*decoder).unmarshal(0xc0009f5110, 0xc00256e460, {0x2998d40?, 0xc00280c018?, 0xc0027de910?})
        /home/runner/go/pkg/mod/gopkg.in/yaml.v3@v3.0.1/decode.go:502 +0x2b2
gopkg.in/yaml%2ev3.(*decoder).mapping(0xc0009f5110, 0xc00256e1e0, {0x262c280?, 0xc0027b63f0?, 0xc0009272e8?})
        /home/runner/go/pkg/mod/gopkg.in/yaml.v3@v3.0.1/decode.go:848 +0x98e
gopkg.in/yaml%2ev3.(*decoder).unmarshal(0xc0009f5110, 0xc00256e1e0, {0x24e3dc0?, 0xc0027b63f0?, 0x7f1b88648d28?})
        /home/runner/go/pkg/mod/gopkg.in/yaml.v3@v3.0.1/decode.go:510 +0x3ea
gopkg.in/yaml%2ev3.(*decoder).callObsoleteUnmarshaler.func1({0x24e3dc0?, 0xc0027b63f0?})
        /home/runner/go/pkg/mod/gopkg.in/yaml.v3@v3.0.1/decode.go:381 +0xe5
github.com/pulumi/pulumi/sdk/v3/go/common/resource/config.(*Map).UnmarshalYAML(0xc0022a91d0, 0xc0027b8520)
        /home/runner/work/pulumi/pulumi/sdk/go/common/resource/config/map.go:239 +0x82
gopkg.in/yaml%2ev3.(*decoder).callObsoleteUnmarshaler(0xc0009f5110, 0xc00256e1e0, {0x7f1b41562068, 0xc0022a91d0})
        /home/runner/go/pkg/mod/gopkg.in/yaml.v3@v3.0.1/decode.go:379 +0x94
gopkg.in/yaml%2ev3.(*decoder).prepare(0x251a320?, 0xc00256e140?, {0x29251e0?, 0xc0022a91d0?, 0x251a320?})
        /home/runner/go/pkg/mod/gopkg.in/yaml.v3@v3.0.1/decode.go:427 +0x21e
gopkg.in/yaml%2ev3.(*decoder).unmarshal(0xc0009f5110, 0xc00256e1e0, {0x29251e0?, 0xc0022a91d0?, 0x9f980f?})
        /home/runner/go/pkg/mod/gopkg.in/yaml.v3@v3.0.1/decode.go:502 +0x2b2
gopkg.in/yaml%2ev3.(*decoder).mappingStruct(0xc0009f5110, 0xc00256e0a0, {0x2952ae0?, 0xc0022a91a0?, 0x40dc5f?})
        /home/runner/go/pkg/mod/gopkg.in/yaml.v3@v3.0.1/decode.go:935 +0xf74
gopkg.in/yaml%2ev3.(*decoder).mapping(0xc0009f5110, 0xc00256e0a0, {0x2952ae0?, 0xc0022a91a0?, 0x0?})
        /home/runner/go/pkg/mod/gopkg.in/yaml.v3@v3.0.1/decode.go:786 +0xa5
gopkg.in/yaml%2ev3.(*decoder).unmarshal(0xc0009f5110, 0xc00256e0a0, {0x2952ae0?, 0xc0022a91a0?, 0x9fa8a5?})
        /home/runner/go/pkg/mod/gopkg.in/yaml.v3@v3.0.1/decode.go:510 +0x3ea
gopkg.in/yaml%2ev3.(*decoder).document(...)
        /home/runner/go/pkg/mod/gopkg.in/yaml.v3@v3.0.1/decode.go:527
gopkg.in/yaml%2ev3.(*decoder).unmarshal(0xc0009f5110, 0xc00256e000, {0x2952ae0?, 0xc0022a91a0?, 0x0?})
        /home/runner/go/pkg/mod/gopkg.in/yaml.v3@v3.0.1/decode.go:498 +0x28a
gopkg.in/yaml%2ev3.unmarshal({0xc002470000, 0xd5e5, 0x10000}, {0x26f92a0, 0xc0022a91a0}, 0x58?)
        /home/runner/go/pkg/mod/gopkg.in/yaml.v3@v3.0.1/yaml.go:167 +0x3b1
gopkg.in/yaml%2ev3.Unmarshal(...)
        /home/runner/go/pkg/mod/gopkg.in/yaml.v3@v3.0.1/yaml.go:89
github.com/pulumi/pulumi/sdk/v3/go/common/encoding.(*yamlMarshaler).Unmarshal(0xc0021f4370?, {0xc002470000?, 0xc002316e70?, 0x47eff60?}, {0x26f92a0?, 0xc0022a91a0?})
        /home/runner/work/pulumi/pulumi/sdk/go/common/encoding/marshal.go:113 +0x39
github.com/pulumi/pulumi/sdk/v3/go/common/workspace.LoadProjectStack(0xc0020edb78?, {0xc002087c00, 0x3d})
        /home/runner/work/pulumi/pulumi/sdk/go/common/workspace/loaders.go:192 +0x1f9
github.com/pulumi/pulumi/sdk/v3/go/common/workspace.DetectProjectStack({0xc0020edb78?, 0x2a66e83?})
        /home/runner/work/pulumi/pulumi/sdk/go/common/workspace/paths.go:162 +0x30
main.loadProjectStack(0x2a5d883?, {0x329b3e0?, 0xc0020bf170?})
        /home/runner/work/pulumi/pulumi/pkg/cmd/pulumi/config.go:738 +0x3c
main.getStackConfigurationWithFallback({0x32830d0, 0x4874ea0}, {0x329b3e0, 0xc0020bf170}, 0x32830d0?, {0x0, 0x0}, 0x0)
        /home/runner/work/pulumi/pulumi/pkg/cmd/pulumi/config.go:1113 +0xa5
main.getStackConfiguration(...)
        /home/runner/work/pulumi/pulumi/pkg/cmd/pulumi/config.go:1024
main.newUpCmd.func1({_, _}, {{{0x0, 0x0, 0x0}, {0x0, 0x0, 0x0}, 0x0, 0x0, ...}, ...}, ...)
        /home/runner/work/pulumi/pulumi/pkg/cmd/pulumi/up.go:109 +0x309
main.newUpCmd.func3(0x0?, {0x4874ea0, 0x0, 0x0?})
        /home/runner/work/pulumi/pulumi/pkg/cmd/pulumi/up.go:513 +0xb8e
main.newUpCmd.RunResultFunc.func4(0xc000bb0100?, {0x4874ea0, 0x0, 0x0})
        /home/runner/work/pulumi/pulumi/sdk/go/common/util/cmdutil/exit.go:115 +0x38
github.com/spf13/cobra.(*Command).execute(0xc000bae600, {0x4874ea0, 0x0, 0x0})
        /home/runner/go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:944 +0x863
github.com/spf13/cobra.(*Command).ExecuteC(0xc001a51200)
        /home/runner/go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:1068 +0x3a5
github.com/spf13/cobra.(*Command).Execute(...)
        /home/runner/go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:992
main.main()
        /home/runner/work/pulumi/pulumi/pkg/cmd/pulumi/main.go:56 +0x58
``` 

### Output of `pulumi about`

```
Backend        
Name           <backend-name>
URL            <s3-backend>
User           <user>
Organizations  
Token type     personal

Dependencies:
NAME        VERSION
boto3       1.28.53
fabric      3.2.2
pip         22.3.1
psutil      5.9.5
pulumi-aws  6.4.0
requests    2.31.0
setuptools  65.5.1
wheel       0.38.4

Pulumi locates its logs in /tmp by default

``` 

### Additional context

_No response_

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

### PR DESCRIPTION
**Pull request by tgummerer**
Before #13814, when we encountered unexpected types in a yaml file, we ignored them.  However that PR changed that, and we end up panicing, e.g. in case we have a config like the following:

```
config:
  aws:region: us-east-1
  aws:id:
    -
```

Fix that by getting the old behaviour of ignoring this error back.

Fixes #14146 

## Checklist

- [x] I have run `make tidy` to update any new dependencies
- [x] I have run `make lint` to verify my code passes the lint check
  - [x] I have formatted my code using `gofumpt`

- [x] I have added tests that prove my fix is effective or that my feature works

- [x] I have run `make changelog` and committed the `changelog/pending/<file>` documenting my change

- [ ] Yes, there are changes in this PR that warrants bumping the Pulumi Cloud API version

### CODE REVIEW

**Review by Zaid-Ajaj — APPROVED** [review]
LGTM

**Comment by Frassle** [comment]
A) Can we have a test to cover this.
B) I am uneasy that we just ignore these errors, why is this an unrecognized type? It looks like it should match against `[]object`?

**Comment by tgummerer** [comment]
> A) Can we have a test to cover this.

I'm looking into that, but I'm struggling a little bit with getting it right.  @pgavlin added these expectation tests which seem nice, but I haven't quite figured out how to have the right base64 encoded object there. (https://github.com/pulumi/pulumi/pull/13814/files#diff-79d7d86f6e3d374cae7636670dba41effb46c28701c3b04a0b0634689d585c03)

> B) I am uneasy that we just ignore these errors, why is this an unrecognized type? It looks like it should match against `[]object`?

It's `nil` and not `[]object` for some reason that I can't claim to fully understand.  Maybe we match against that instead?

File: tests/integration/integration_test.go [hunk 1]
```diff
@@ -494,7 +494,6 @@ func TestConfigPaths(t *testing.T) {
 		"root[",
 		`root["nested]`,
 		"root.array[abc]",
-		"root.[1]",
```
Comment (dixler) [inline tests/integration/integration_test.go:497]: [null, "value"]

**Comment by justinvp** [comment]
@dixler, why don't we have to handle `nil` in `decrypt`?

**Comment by dixler** [comment]
> @dixler, why don't we have to handle `nil` in `decrypt`?

I wasn't sure why we did in the first place as it wasn't noted in the panics that were reported and I was trying to minimize the number of changes when there was no background I was aware of.

**Comment by tgummerer** [comment]
> I wasn't sure why we did in the first place as it wasn't noted in the panics that were reported and I was trying to minimize the number of changes when there was no background I was aware of.

It was there because I noticed it crashes in that location after fixing the other issue:

<details>
<summary>Here's the stack trace:</summary>

```
================================================================================
The Pulumi CLI encountered a fatal error. This is a bug!
We would appreciate a report: https://github.com/pulumi/pulumi/issues/
Please provide all of the below text in your report.
================================================================================
Pulumi Version:   3.88.1-dev.0
Go Version:       go1.20.8
Go Compiler:      gc
Architecture:     amd64
Operating System: linux
Panic:            fatal: A failure has occurred: unexpected value of type <nil>

goroutine 1 [running]:
runtime/debug.Stack()
	/nix/store/23wrmhyfnwjxc9cvy3lab9xj551ic53c-go-1.20.8/share/go/src/runtime/debug/stack.go:24 +0x65
main.panicHandler(0xc001e2bf47)
	/home/tgummerer/work/pulumi/main/pkg/cmd/pulumi/main.go:35 +0x47
panic({0x2478560, 0xc00052c410})
	/nix/store/23wrmhyfnwjxc9cvy3lab9xj551ic53c-go-1.20.8/share/go/src/runtime/panic.go:884 +0x213
github.com/pulumi/pulumi/sdk/v3/go/common/util/contract.failfast(...)
	/home/tgummerer/work/pulumi/main/sdk/go/common/util/contract/failfast.go:23
github.com/pulumi/pulumi/sdk/v3/go/common/util/contract.Failf({0x2a2c6e7?, 0x40f94a?}, {0xc001e29a18?, 0x18?, 0xc001e29a18?})
	/home/tgummerer/work/pulumi/main/sdk/go/common/util/contract/fail.go:32 +0xcd
github.com/pulumi/pulumi/sdk/v3/go/common/resource/config.object.decrypt({{0x0?, 0x0?}, 0x20?}, {0x31126a0?, 0xc000058108?}, {0xc00052c3e0?, 0xc00054a168?, 0x241e120?}, {0x7f54dc2525f8, 0x463a928})
	/home/tgummerer/work/pulumi/main/sdk/go/common/resource/config/object.go:119 +0x3c5
github.com/pulumi/pulumi/sdk/v3/go/common/resource/config.object.decrypt({{0x243c7c0?, 0xc000538348?}, 0x1a?}, {0x31126a0, 0xc000058108}, {0x0?, 0x0, 0x0}, {0x7f54dc2525f8, 0x463a928})
	/home/tgummerer/work/pulumi/main/sdk/go/common/resource/config/object.go:101 +0x531
github.com/pulumi/pulumi/sdk/v3/go/common/resource/config.object.Decrypt(...)
	/home/tgummerer/work/pulumi/main/sdk/go/common/resource/config/object.go:78
github.com/pulumi/pulumi/sdk/v3/go/common/resource/config.Value.Value({{0xc0024ebc50?, 0xc0005366f0?}, 0x40?, 0xa4?}, {0x7f54dc2525f8?, 0x463a928})
	/home/tgummerer/work/pulumi/main/sdk/go/common/resource/config/value.go:55 +0xc5
github.com/pulumi/pulumi/sdk/v3/go/common/resource/config.Map.Decrypt(0x7f55063c1a68?, {0x7f54dc2525f8, 0x463a928})
	/home/tgummerer/work/pulumi/main/sdk/go/common/resource/config/map.go:33 +0xf5
github.com/pulumi/pulumi/pkg/v3/engine.newDeployment(0xc00053e1c0, 0xc00055a140, 0xc000562000, 0x78?)
	/home/tgummerer/work/pulumi/main/pkg/engine/deployment.go:175 +0x1b1
github.com/pulumi/pulumi/pkg/v3/engine.update(0xc00053e1c0, 0xc00055a140, 0xc000562000, 0x1)
	/home/tgummerer/work/pulumi/main/pkg/engine/update.go:464 +0x217
github.com/pulumi/pulumi/pkg/v3/engine.Update({0x310cd90, 0xc00054c870}, 0xc00053e1c0, {{0x463a928, 0x0, 0x0}, {0x0, 0x0, 0x0}, 0x7fffffff, ...}, ...)
	/home/tgummerer/work/pulumi/main/pkg/engine/update.go:212 +0x525
github.com/pulumi/pulumi/pkg/v3/backend/httpstate.(*cloudBackend).runEngineAction(_, {_, _}, {_, _}, {_, _}, {0xc0005b4bb0, {0xc00248dfb0, 0x23}, ...}, ...)
	/home/tgummerer/work/pulumi/main/pkg/backend/httpstate/backend.go:1269 +0x931
github.com/pulumi/pulumi/pkg/v3/backend/httpstate.(*cloudBackend).apply(_, {_, _}, {_, _}, {_, _}, {0xc0005b4bb0, {0xc00248dfb0, 0x23}, ...}, ...)
	/home/tgummerer/work/pulumi/main/pkg/backend/httpstate/backend.go:1194 +0x568
github.com/pulumi/pulumi/pkg/v3/backend/httpstate.(*cloudBackend).Preview(_, {_, _}, {_, _}, {0xc0005b4bb0, {0xc00248dfb0, 0x23}, {0x0, 0x0, ...}, ...})
	/home/tgummerer/work/pulumi/main/pkg/backend/httpstate/backend.go:1013 +0xa7
github.com/pulumi/pulumi/pkg/v3/backend.PreviewStack({_, _}, {_, _}, {0xc0005b4bb0, {0xc00248dfb0, 0x23}, {0x0, 0x0, 0x0}, ...})
	/home/tgummerer/work/pulumi/main/pkg/backend/stack.go:89 +0xba
github.com/pulumi/pulumi/pkg/v3/backend/httpstate.(*cloudStack).Preview(_, {_, _}, {0xc0005b4bb0, {0xc00248dfb0, 0x23}, {0x0, 0x0, 0x0}, 0xc0024f0bb8, ...})
	/home/tgummerer/work/pulumi/main/pkg/backend/httpstate/stack.go:162 +0x88
main.newPreviewCmd.func1(0x0?, {0x463a928, 0x0, 0x0?})
	/home/tgummerer/work/pulumi/main/pkg/cmd/pulumi/preview.go:239 +0x1670
github.com/pulumi/pulumi/sdk/v3/go/common/util/cmdutil.RunResultFunc.func1(0xc00104c300?, {0x463a928, 0x0, 0x0})
	/home/tgummerer/work/pulumi/main/sdk/go/common/util/cmdutil/exit.go:115 +0x3e
github.com/spf13/cobra.(*Command).execute(0xc00104c300, {0x463a928, 0x0, 0x0})
	/home/tgummerer/go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:944 +0x847
github.com/spf13/cobra.(*Command).ExecuteC(0xc00092c000)
	/home/tgummerer/go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:1068 +0x3bd
github.com/spf13/cobra.(*Command).Execute(...)
	/home/tgummerer/go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:992
main.main()
	/home/tgummerer/work/pulumi/main/pkg/cmd/pulumi/main.go:56 +0x65
```

</details>

**Comment by tgummerer** [comment]
I added a commit fixing that again, and adding to the test @dixler added to cover that case.

File: sdk/go/common/resource/config/object.go [hunk 2]
```diff
@@ -115,6 +116,8 @@ func (c object) decrypt(ctx context.Context, path resource.PropertyPath, decrypt
 			vs[k] = pv
 		}
 		return NewPlaintext(vs), nil
+	case nil:
+		return Plaintext{}, nil
```
Comment (dixler) [inline sdk/go/common/resource/config/object.go:120]: Can you add a test that exercises this part of the code?

Comment (tgummerer) [reply sdk/go/common/resource/config/object.go:120]: There is.  https://github.com/pulumi/pulumi/pull/14149/files#diff-2ee802e4d2ddbf1ae1787d72e96c45f45bfa2586c4dab5dc9f663d1705f7bc66R40-R41 should exercise this, and panics if we're missing this piece here.

**Review by dixler — APPROVED** [review]

**Pull request merged**